File: exemples/js/05_unregister_operator.js
/**
 * Example: Unregister an operator
 *
 * This example shows how to:
 * 1. Check current operator registration status
 * 2. Unregister an operator
 * 3. Verify the unregistration
 *
 * WARNING: After unregistering, the operator cannot process new payments
 * until they register again.
 */

import {
  initTronWeb,
  getPaymentSplitterContract,
  waitForTransaction,
  printResult
} from './utils.js';

async function main() {
  try {
    console.log('=== Unregister Operator Example ===\n');

    // Initialize TronWeb (using default account from .env)
    const tronWeb = initTronWeb();
    console.log('Connected to Shasta testnet');
    console.log(`Operator address: ${tronWeb.defaultAddress.base58}\n`);

    // Load PaymentSplitter contract
    const splitter = await getPaymentSplitterContract(tronWeb);
    console.log('PaymentSplitter contract loaded\n');

    // Check current registration status
    console.log('Checking current registration status...');
    const isRegisteredBefore = await splitter.isOperatorRegistered(tronWeb.defaultAddress.base58).call();

    if (!isRegisteredBefore) {
      console.log('✗ Operator is not registered. Nothing to unregister.\n');
      return;
    }

    const feeDestinationBefore = await splitter.getFeeDestination(tronWeb.defaultAddress.base58).call();
    console.log(`Current status: Registered`);
    console.log(`Fee destination: ${tronWeb.address.fromHex(feeDestinationBefore)}\n`);

    // Unregister operator
    console.log('Unregistering operator...');
    const tx = await splitter.unregisterOperator().send({
      feeLimit: 100_000_000,
      callValue: 0
    });

    // Wait for confirmation
    await waitForTransaction(tronWeb, tx);

    // Verify unregistration
    console.log('\nVerifying unregistration...');
    const isRegisteredAfter = await splitter.isOperatorRegistered(tronWeb.defaultAddress.base58).call();
    const feeDestinationAfter = await splitter.getFeeDestination(tronWeb.defaultAddress.base58).call();

    printResult('Unregistration Result', {
      'Operator': tronWeb.defaultAddress.base58,
      'Was Registered': isRegisteredBefore,
      'Is Registered Now': isRegisteredAfter,
      'Fee Destination': tronWeb.address.fromHex(feeDestinationAfter),
      'Transaction': tx
    });

    console.log('✓ Operator unregistered successfully!\n');
    console.log('⚠️  Note: The operator can no longer process payments until re-registered.\n');

  } catch (error) {
    console.error('Error:', error.message);
    process.exit(1);
  }
}

// Run the example
main();

File: exemples/ruby/05_unregister_operator.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# Example: Unregister an operator
#
# This example shows how to:
# 1. Check current operator registration status
# 2. Unregister an operator
# 3. Verify the unregistration
#
# WARNING: After unregistering, the operator cannot process new payments
# until they register again.
#
# IMPORTANT: Run with bundle exec to avoid gem conflicts:
#   cd examples && bundle exec ruby ruby/05_unregister_operator.rb

require_relative 'utils'

def main
  puts '=== Unregister Operator Example ==='
  puts

  # Initialize Tron client (using default account from .env)
  client = init_tron_client
  puts 'Connected to Shasta testnet'

  operator_private_key = ENV['OPERATOR_PRIVATE_KEY']
  unless operator_private_key
    puts '❌ Error: OPERATOR_PRIVATE_KEY not found in .env file'
    exit 1
  end

  operator_address = get_address_from_private_key(operator_private_key)
  puts "Operator address: #{operator_address}"
  puts

  begin
    contract_service = client.contract_service

    # Check current registration status
    puts 'Checking current registration status...'
    is_registered_before = contract_service.operator_registered?(
      PAYMENT_SPLITTER_ADDRESS,
      operator_address
    )

    unless is_registered_before
      puts '✗ Operator is not registered. Nothing to unregister.'
      puts
      return
    end

    fee_destination_before = contract_service.get_fee_destination(
      PAYMENT_SPLITTER_ADDRESS,
      operator_address
    )

    puts 'Current status: Registered'
    puts "Fee destination: #{fee_destination_before}"
    puts

    # Unregister operator
    puts 'Unregistering operator...'
    tx = contract_service.trigger_contract(
      contract_address: PAYMENT_SPLITTER_ADDRESS,
      function: 'unregisterOperator()',
      parameters: [],
      private_key: operator_private_key
    )

    tx_id = tx['txid'] || tx['txID']
    wait_for_transaction(client, tx_id)

    # Verify unregistration
    puts
    puts 'Verifying unregistration...'
    is_registered_after = contract_service.operator_registered?(
      PAYMENT_SPLITTER_ADDRESS,
      operator_address
    )

    fee_destination_after = contract_service.get_fee_destination(
      PAYMENT_SPLITTER_ADDRESS,
      operator_address
    )

    print_result('Unregistration Result', {
      'Operator' => operator_address,
      'Was Registered' => is_registered_before,
      'Is Registered Now' => is_registered_after,
      'Fee Destination' => fee_destination_after,
      'Transaction' => tx_id
    })

    puts '✓ Operator unregistered successfully!'
    puts
    puts '⚠️  Note: The operator can no longer process payments until re-registered.'
    puts

  rescue => e
    puts "❌ Error: #{e.message}"
    puts "Backtrace: #{e.backtrace.first(3).join("\n")}" if ENV['DEBUG']
    exit 1
  end
end

main if __FILE__ == $PROGRAM_NAME
